### ISSUE TITLE
`foldBy` and `foldMapBy` should probably be moved elsewhere

### ISSUE THREAD

**Issue by treeowl** [ekmett/lens#577]
Could these fit in `reflection-extras` or some such? Let me know if you want me to do some of the grunt work on that.

**Comment by ekmett** [comment]
I'm open to moving them up into `reflection` and then leaving them re-exported where they are exported now. 

This isn't without (temporary) cost, though, as it marches forward the floor of the version of `reflection` we are compatible with, rendering lots of build plans inaccessible.

As we discussed at Hac Boston, I don't think `reflection-extras` is a good home for them, though.

**Comment by treeowl** [comment]
I don't see the problem. First add `Data.Reflection.Monoid`; wait a couple cycles before using it in `lens`. Might want to add reflective versions of `Foldable`, `Traversable`, and `Num` (for Oleg's classic modular arithmetic example) while you're at it. Let me know if you need a code monkey to help.

**Comment by treeowl** [comment]
Actually, I want to give the `Foldable` one a shot. I have some thoughts about implementing defaulting.

**Comment by ekmett** [comment]
I'd rather export both combinators from `Data.Reflection`. They are innocuous enough.

We can then put out a version of `lens` which keeps the current bounds on `reflection`, but re-exports them if it is on a new enough version, and eventually raise the bound on `lens`.

Whether the version bump on `reflection` is high enough to skip past the current `lens` range or we do revisions I'm somewhat neutral on. The former would require a first digit bump for a fairly minor addition. The latter requires a lot of work.

That maximizes the number of build plans and avoids randomly shoving these things in an undiscoverable module in `reflection`.

**Comment by ekmett** [comment]
`traverseBy` would be pretty nice to have, yeah.

**Comment by ekmett** [comment]
Ok, we'd only have to do revisions on `lens` 4.12+ so I think the new reflection version should be 2.1 to contain these additions.

I'm thinking we may want to leave Num off of the list as it is a rather large mess to work with.

**Comment by ekmett** [comment]
I've made the change in HEAD, but I haven't yet shipped `reflection` 2.1. Before we can release `lens` 4.13 we'll need to release `reflection` 2.1.

**Comment by treeowl** [comment]
Num is bad, but we can handle it, I think. Take a record with some Maybes
and build a dictionary dealing with defaults by hand. Building the rest of
the numeric hierarchy could be interesting.

**Comment by ekmett** [comment]
There is a difference between can and should. =)

This sounds like a good candidate to exile into `reflection-extras` where it can be silently forgotten. =)
